javacord-core/src/main/java/org/javacord/core/DiscordApiImpl.java

```diff
@@ -1344,117 +1351,212 @@ public ThreadPool getThreadPool() {
+            permissions.add(new ServerApplicationCommandPermissionsImpl(jsonNode));
         }
         return permissions;
     }
 
+    private List<ApplicationCommand> jsonToApplicationCommandList(JsonNode resultJson) {
+        List<ApplicationCommand> applicationCommands = new ArrayList<>();
+        for (JsonNode applicationCommandJson : resultJson) {
+            CommandType type = CommandType.fromValue(applicationCommandJson.get("type").asInt());
+            if (type.equals(CommandType.SLASH)) {
+                applicationCommands.add(new SlashCommandImpl(this, applicationCommandJson));
+            }
+            if (type.equals(CommandType.USER)) {
+                applicationCommands.add(new UserContextMenuImpl(this, applicationCommandJson));
+            }
+            if (type.equals(CommandType.MESSAGE)) {
+                applicationCommands.add(new MessageContextMenuImpl(this, applicationCommandJson));
+            }
+        }
+        return Collections.unmodifiableList(applicationCommands);
+    }
+
     private List<SlashCommand> jsonToSlashCommandList(JsonNode resultJson) {
         List<SlashCommand> slashCommands = new ArrayList<>();
         for (JsonNode slashCommandJson : resultJson) {
-            slashCommands.add(new SlashCommandImpl(this, slashCommandJson));
+            if (CommandType.fromValue(slashCommandJson.get("type").asInt()).equals(CommandType.SLASH)) {
+                slashCommands.add(new SlashCommandImpl(this, slashCommandJson));
+            }
         }
         return Collections.unmodifiableList(slashCommands);
     }
 
+    private List<UserContextMenu> jsonToUserContextMenuList(JsonNode resultJson) {
+        List<UserContextMenu> userContextMenus = new ArrayList<>();
+        for (JsonNode userContextMenuJson : resultJson) {
+            if (CommandType.fromValue(userContextMenuJson.get("type").asInt()).equals(CommandType.USER)) {
+                userContextMenus.add(new UserContextMenuImpl(this, userContextMenuJson));
+            }
+        }
+        return Collections.unmodifiableList(userContextMenus);
+    }
+
+    private List<MessageContextMenu> jsonToMessageContextMenuList(JsonNode resultJson) {
+        List<MessageContextMenu> messageContextMenus = new ArrayList<>();
+        for (JsonNode messageContextMenuJson : resultJson) {
+            if (CommandType.fromValue(messageContextMenuJson.get("type").asInt()).equals(CommandType.MESSAGE)) {
+                messageContextMenus.add(new MessageContextMenuImpl(this, messageContextMenuJson));
+            }
+        }
+        return Collections.unmodifiableList(messageContextMenus);
+    }
+
     @Override
-    public CompletableFuture<List<SlashCommand>> bulkOverwriteGlobalSlashCommands(
-            List<SlashCommandBuilder> slashCommandBuilderList) {
+    public CompletableFuture<List<ApplicationCommand>> bulkOverwriteGlobalApplicationCommands(
+            List<ApplicationCommandBuilder<?>> applicationCommandBuilderList) {
```

Comment:
This probably should be changed to 
```suggestion
            List<ApplicationCommandBuilder<? extends ApplicationCommand>> applicationCommandBuilderList) {
```